File: packages/ios/Sources/Components/Itinerary/Itinerary.swift
import SwiftUI

/// One step of an itinerary, as the products present it: the SDK's own
/// wording, the arrow it gets, and whether it is the step under way.
public struct KozmosItineraryStep: Identifiable, Hashable, Sendable {
    public let id: String
    public let instruction: String
    public let type: DirectionType
    public let isCurrent: Bool

    public init(id: String, instruction: String, type: DirectionType, isCurrent: Bool = false) {
        self.id = id
        self.instruction = instruction
        self.type = type
        self.isCurrent = isCurrent
    }
}

/// The whole route as a list: where it starts, every step with the current one
/// emphasised, where it ends. Mirrors the product prototype's itinerary —
/// FROM and TO in 11-point capitals beside their names, steps in 15 with the
/// current one semibold in the theme colour.
///
/// VoiceOver hears the endpoints as "From, name" and "To, name", each step as
/// one element, the current one selected.
public struct KozmosItinerary: View {
    let origin: String
    let steps: [KozmosItineraryStep]
    let destination: String
    let originLabel: String
    let destinationLabel: String
    let label: String

    public init(
        origin: String,
        steps: [KozmosItineraryStep],
        destination: String,
        originLabel: String = "From",
        destinationLabel: String = "To",
        label: String = "Itinerary"
    ) {
        self.origin = origin
        self.steps = steps
        self.destination = destination
        self.originLabel = originLabel
        self.destinationLabel = destinationLabel
        self.label = label
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: KozmosDimensions.primitivesLayoutSpacing100) {
            endpoint(originLabel, name: origin, emphasised: false)
            ForEach(steps) { step in
                row(step)
            }
            endpoint(destinationLabel, name: destination, emphasised: true)
        }
        .accessibilityElement(children: .contain)
        .accessibilityLabel(label)
    }

    private func endpoint(_ label: String, name: String, emphasised: Bool) -> some View {
        HStack(alignment: .firstTextBaseline, spacing: KozmosDimensions.primitivesLayoutSpacing150) {
            Text(label.uppercased())
                .font(KozmosTypography.caption2)
                .foregroundColor(KozmosColors.primitivesColorsForeground500)
                .frame(width: KozmosDimensions.primitivesLayoutSizing500, alignment: .leading)
            Text(name)
                .font(emphasised ? KozmosTypography.subheadline.weight(.semibold) : KozmosTypography.subheadline)
                .foregroundColor(emphasised ? KozmosColors.primitivesColorsForeground100 : KozmosColors.primitivesColorsForeground500)
                .lineLimit(2)
                .fixedSize(horizontal: false, vertical: true)
        }
        .accessibilityElement(children: .ignore)
        .accessibilityLabel("\(label), \(name)")
    }

    private func row(_ step: KozmosItineraryStep) -> some View {
        HStack(alignment: .top, spacing: KozmosDimensions.primitivesLayoutSpacing150) {
            Image(systemName: step.type.iconName)
                .font(.system(size: 14, weight: .semibold))
                .foregroundColor(step.isCurrent ? KozmosColors.primitivesColorsTheme500 : KozmosColors.primitivesColorsForeground500)
                .frame(width: KozmosDimensions.primitivesLayoutSizing500, height: 20, alignment: .leading)
            Text(step.instruction)
                .font(step.isCurrent ? KozmosTypography.subheadline.weight(.semibold) : KozmosTypography.subheadline)
                .foregroundColor(step.isCurrent ? KozmosColors.primitivesColorsTheme500 : KozmosColors.primitivesColorsForeground100)
                .fixedSize(horizontal: false, vertical: true)
        }
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(step.instruction)
        .accessibilityAddTraits(step.isCurrent ? .isSelected : [])
    }
}

File: packages/android/src/main/java/com/kozmos/components/Itinerary/Itinerary.kt
package com.kozmos.components.itinerary

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.kozmos.components.directionstep.DirectionType
import com.kozmos.components.directionstep.icon
import com.kozmos.tokens.KozmosColors
import com.kozmos.tokens.KozmosDimensions

/** One step of an itinerary, as the products present it. */
data class KozmosItineraryStep(
    val id: String,
    /** The routing engine's own wording. */
    val instruction: String,
    val type: DirectionType,
    /** The step under way. */
    val isCurrent: Boolean = false
)

/**
 * The whole route as a list: where it starts, every step with the current
 * one emphasised, where it ends. TalkBack hears the endpoints as "From,
 * name" and "To, name", each step as one element, the current one selected.
 */
@Composable
fun KozmosItinerary(
    origin: String,
    steps: List<KozmosItineraryStep>,
    destination: String,
    modifier: Modifier = Modifier,
    originLabel: String = "From",
    destinationLabel: String = "To",
    label: String = "Itinerary"
) {
    Column(
        modifier = modifier
            .fillMaxWidth()
            .semantics { contentDescription = label },
        verticalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing100)
    ) {
        Endpoint(originLabel, origin, emphasised = false)
        steps.forEach { step -> StepRow(step) }
        Endpoint(destinationLabel, destination, emphasised = true)
    }
}

@Composable
private fun Endpoint(label: String, name: String, emphasised: Boolean) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .semantics(mergeDescendants = true) { contentDescription = "$label, $name" },
        verticalAlignment = Alignment.Top
    ) {
        Text(
            text = label.uppercase(),
            style = MaterialTheme.typography.labelSmall,
            color = KozmosColors.primitivesColorsForeground500,
            modifier = Modifier.width(KozmosDimensions.primitivesLayoutSizing500).padding(top = 3.dp)
        )
        Text(
            text = name,
            style = MaterialTheme.typography.bodyLarge.copy(fontWeight = if (emphasised) FontWeight.SemiBold else FontWeight.Normal),
            color = if (emphasised) KozmosColors.primitivesColorsForeground100 else KozmosColors.primitivesColorsForeground500,
            modifier = Modifier.padding(start = KozmosDimensions.primitivesLayoutSpacing150)
        )
    }
}

@Composable
private fun StepRow(step: KozmosItineraryStep) {
    val colour = if (step.isCurrent) KozmosColors.primitivesColorsTheme500 else KozmosColors.primitivesColorsForeground100
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .semantics(mergeDescendants = true) {
                contentDescription = step.instruction
                selected = step.isCurrent
            },
        verticalAlignment = Alignment.Top
    ) {
        Icon(
            imageVector = step.type.icon(),
            contentDescription = null,
            tint = if (step.isCurrent) KozmosColors.primitivesColorsTheme500 else KozmosColors.primitivesColorsForeground500,
            modifier = Modifier
                .width(KozmosDimensions.primitivesLayoutSizing500)
                .height(20.dp)
                .padding(end = KozmosDimensions.primitivesLayoutSpacing300)
        )
        Text(
            text = step.instruction,
            style = MaterialTheme.typography.bodyLarge.copy(fontWeight = if (step.isCurrent) FontWeight.SemiBold else FontWeight.Normal),
            color = colour,
            modifier = Modifier.padding(start = KozmosDimensions.primitivesLayoutSpacing150)
        )
    }
}
